Pr/0002 ficr service in spm

pkchan opened the pull request:
There are applications which requires reading of FICR. Therefore, there needs to be an interface to read them.

Range 0x204 ~ 0xC20 are published in infocenter.
Range 0x130 ~ 0x134 happens in system_nrf9160.c. They are used to check whether certain errata fixes are required. It does not harm to read them.

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## Gitlint issues

Commit 49d69fdf04:
1: UC6 Body has no content, should at least have 1 line.

Commit d79bc8682f:
1: UC6 Body has no content, should at least have 1 line.

Commit f7815723e1:
1: UC6 Body has no content, should at least have 1 line.

Commit 5d589c6be6:
1: UC6 Body has no content, should at least have 1 line.

pizi-nordic commented on subsys/spm/secure_services.c:
```diff
@@ -52,6 +52,23 @@ int spm_secure_services_init(void)
+		return -EINVAL;
+	}
+
+	if ((offset >= 0x204 && offset < 0xC20) ||
```
Could we use some defines for all constants here?

pkchan replied:
Yes, we could.

pkchan commented on subsys/spm/secure_services.c:
```diff
@@ -52,6 +52,23 @@ int spm_secure_services_init(void)
+	}
+
+	if ((offset >= 0x204 && offset < 0xC20) ||
+		(offset >= 0x130 && offset <= 0x134)) {
```
This has the same purpose as the functions which checks FICR in system_nrf9160.c (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/1e34490406b4231fa79558d2759264397fbefa4c/ext/hal/nordic/nrfx/mdk/system_nrf9160.c#L183)). Reading the 4-byte register with offset 0x134 is intentional. If it is OK in system_nrf9160.c, then it should also be OK here.

u32_t * already implies that this should be 4-byte aligned. If the user is providing something odd, the final behavior will be governed by Coretex-M33.

You reminded me of one more thing: 0xC20 should be changed to 0xC1C. This is to make sure access to anything beyond 0xC20 will be discarded. This will be scheduled to the next rework.

pkchan commented:
So far, according to the comments collected in the past two days, I have come up with a proposal. If it is OK, then a rework will be scheduled in the next office day.

**Proposal:**
1. Only two commits: one for the feature, the other for the example (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1172#pullrequestreview-291609644));
2. Change 0xC20 to 0xC1C (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1172#discussion_r327950057));
3. Use defines for constants 0x204, 0xC1C, 0x130 and 0x134 (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1172#discussion_r327081902));

pkchan commented:
Reworked according to proposal (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1172#issuecomment-534876791)) and then uploaded.

Rebased to master d9b7f77 (as of 4am 26th September, 2019 HKT). No conflict need to be resolved.

ioannisg commented on subsys/spm/secure_services.c:
```diff
@@ -52,6 +52,31 @@ int spm_secure_services_init(void)
+		offset <= FICR_OFFSET_PUBLIC_END) ||
+		(offset >= FICR_OFFSET_RESTRICTED_START &&
+		offset <= FICR_OFFSET_RESTRICTED_END)) {
+		*destination = *(volatile u32_t *)(0x00FF0000 + offset);
```
Can you use some nRF SOC or Dts define for FICR base address

pkchan replied:
Per your questions:
_1. Question: In what way this secure service is different from spm_request_read() which effectively reads some secure memory?_

With spm_request_read() you can only read from the memory starting from PM_MCUBOOT_PAD_ADDRESS to PM_MCUBOOT_PAD_ADDRESS + PM_MCUBOOT_PAD_SIZE. FICR address is outside this range, therefore it is needed. For the reason why FICR need to be accessed from within application, it would be best if we could continue with discussion in private messages.

_2. Can you use some nRF SOC or Dts define for FICR base address?_

In system_nrf9160.c, FICR is accessed by physical address (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/1e34490406b4231fa79558d2759264397fbefa4c/ext/hal/nordic/nrfx/mdk/system_nrf9160.c#L183)). If it is OK there, then it should also be OK here.

Yet, it is a good idea to use macros from nrf9160.h (i.e. NRF_FICR_S_BASE). It is do-able, and will be changed in the next upload.

ioannisg replied:
>With spm_request_read() you can only read from the memory starting from PM_MCUBOOT_PAD_ADDRESS to PM_MCUBOOT_PAD_ADDRESS + PM_MCUBOOT_PAD_SIZE. FICR address is outside this range, therefore it is needed

@pkchan why don't you refactor the existing spm_request_read() to work for any address, offset, and size? After all, it is all about reading some secure memory.

ioannisg replied:
>In system_nrf9160.c, FICR is accessed by physical address (ref: here). If it is OK there, then it should also be OK here.

system_nrf9160.c is a source file from the nRFx project, that is supposed to work, also, outside the context of Zephyr, even, with bare-metal applications. In Zephyr, however, we adopt a policy for having human-readable macros as much as possible. That's why we have DTS to generate peripheral base addresses (in case we do not take them directly from vendor header, e.g. MDK).

pkchan commented:
Changed according to the recommendations from @ioannisg (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1172#discussion_r328599633)), and then uploaded.

Rebased to master and tested functionality using sample project samples\nrf9160\secure_services.

lemrey commented:
> why don't you refactor the existing spm_request_read() to work for any address, offset, and size? After all, it is all about reading some secure memory.

I see no harm in providing a dedicated API to read the FICR register, that way an application can choose which one of the two API they need and disable the other; because we have two APIs that are independently configurable. It could help to limit the amount of information that is exposed, and keep each API as simple as possible. Plus, I see the two API as different, one reading one word from a register, another reading a memory range from flash.

ioannisg commented:
> 
> 
> > why don't you refactor the existing spm_request_read() to work for any address, offset, and size? After all, it is all about reading some secure memory.
> 
> I see no harm in providing a dedicated API to read the FICR register, that way an application can choose which one of the two API they need and disable the other; because we have two APIs that are independently configurable. It could help to limit the amount of information that is exposed, and keep each API as simple as possible. Plus, I see the two API as different, one reading one word from a register, another reading a memory range from flash.

This is not scalable, @lemrey , we need to unify the secure-read operations into one service. Providing two APIs not only wastes flash resources, but also increases the Secure-Non-Secure surface, strictly discouraged by Secure Architecture principles. In the future, one might want to read contents of yet another Secure peripheral (Crypto, KMU, some TIMER, UART, or anything else, or even SRAM); are we going to add  new services for that?

>It could help to limit the amount of information that is exposed

This limiting or exposure of resources is supposed to be a functionality of SPM, not of the API to the NS firmware.

>I see the two API as different, one reading one word from a register, another reading a memory range from flash.

Right, so let's unify them ;)

pkchan commented:
+1 vote here for @ioannisg 's idea. It's just 8 bytes of flash per supported region, if spm_request_read() is reused and if only the ranges[] variable is changed. This is way better than having a new function.

lemrey commented:
I see value in letting an application access only certain addresses/ranges, rather than either all/none. Otherwise all applications that need to read any secure memory will have this functionality enabled and have access to all readable secure memory instead of only the memory that they need. I think that increases the attack surface too.
But if PK is convinced, go ahead :)

pkchan commented:
OK, till now we have 3 votes for @ioannisg 's idea. Let me go for it.

Changes will happen only in spm_request_read(). Since the function has already been demonstrated in samples\nrf9160\secure_services, example codes will be removed from the PR.

For me, the requirement is just to have an interface for our customers to read the FICR without resulting in hardfault. Reading registers from the opened range does not do any harm. Therefore, any method is OK.

ioannisg commented on subsys/spm/secure_services.c:
```diff
@@ -68,6 +76,8 @@ int spm_request_read(void *destination, u32_t addr, size_t len)
 		{.start = PM_MCUBOOT_PAD_ADDRESS,
 		 .size = PM_MCUBOOT_PAD_SIZE},
 #endif
+		{FICR_PUBLIC_ADDR, FICR_PUBLIC_SIZE},
```
can you use the same style for the instantiation here, as above?

pkchan replied:
Yes, will do.

ioannisg commented on subsys/spm/secure_services.c:
```diff
@@ -54,6 +55,13 @@ int spm_secure_services_init(void)
 
 #ifdef CONFIG_SPM_SERVICE_READ
+
+#define FICR_BASE               NRF_FICR_S_BASE
```
Ideally, these are generic memory defines, and could be moved into a soc-specific header, could they?

pkchan replied:
Up to now, secure_service happens only in nRF9160. An extra header file will complicate the situation, so I would rather not recommend. 

A better solution is to rename NRF_FICR_S to NRF_FICR in zephyr\ext\hal\nordic\nrfx\mdk\nrf9160.h. FICR does not have non-secure mapping, so there is no need to explicitly distinguish between secure/non-secure. Since the physical base address is the same as other nRF5 series, using the same name can result in better code reuse.

A simpler solution is to add one more line of #define in zephyr\ext\hal\nordic\nrfx\mdk\nrf9160.h:

`#define NRF_FICR NRF_FICR_S`

This simplifies things.

NordicBuilder commented:
**All checks passed.**

## checkpatch (informational only, not a failure)

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:11: WARNING:LINE_SPACING: Missing a blank line after declarations
#11: FILE: samples/nrf9160/secure_services/src/main.c:70:
+	u32_t ficr_addr = (u32_t)&NRF_FICR_S->INFO.PART;
+	printk("\nRead FICR, offset 0x20C (address 0x%08x):\n", ficr_addr);

- total: 0 errors, 1 warnings, 46 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.

```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

pkchan commented:
Changed accordingly and uploaded:
- Changed style per request (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1172#discussion_r330430038))
- Added example in samples\nrf9160\secure_services to read FICR (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1172#pullrequestreview-296088582))

Rebased on c9c833b, the daily master on 4am 9th October, 2019.
